x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/action/CancelReindexDataStreamTransportAction.java

```diff
@@ -0,0 +1,134 @@
+import org.elasticsearch.xpack.migrate.action.CancelReindexDataStreamAction.Response;
+import org.elasticsearch.xpack.migrate.task.ReindexDataStreamTask;
+
+import java.util.Map;
+import java.util.Optional;
+
+public class CancelReindexDataStreamTransportAction extends HandledTransportAction<Request, Response> {
+    private final ClusterService clusterService;
+    private final TransportService transportService;
+
+    @Inject
+    public CancelReindexDataStreamTransportAction(
+        ClusterService clusterService,
+        TransportService transportService,
+        ActionFilters actionFilters
+    ) {
+        super(CancelReindexDataStreamAction.NAME, transportService, actionFilters, Request::new, EsExecutors.DIRECT_EXECUTOR_SERVICE);
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+    }
+
+    @Override
+    protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        String index = request.getIndex();
+        String persistentTaskId = ReindexDataStreamAction.TASK_ID_PREFIX + index;
+        PersistentTasksCustomMetadata persistentTasksCustomMetadata = clusterService.state()
+            .getMetadata()
+            .custom(PersistentTasksCustomMetadata.TYPE);
+        PersistentTasksCustomMetadata.PersistentTask<?> persistentTask = persistentTasksCustomMetadata.getTask(persistentTaskId);
+        if (persistentTask == null) {
+            listener.onFailure(new ResourceNotFoundException("No migration reindex task found for [{}]", index));
+        } else if (persistentTask.isAssigned()) {
+            String nodeId = persistentTask.getExecutorNode();
+            if (clusterService.localNode().getId().equals(nodeId)) {
+                cancelTaskOnThisNode(persistentTaskId, listener);
+            } else {
+                runActionOnNodeWithTaskIfPossible(task, request, nodeId, listener);
+            }
+        } else {
+            listener.onFailure(new ResourceNotFoundException("Persistent task with id [{}] is not assigned to a node", persistentTaskId));
+        }
+    }
+
+    /*
+     * This method cancels the persistent task with id persistentTaskId. It assumes that the task is running on this node.
+     */
+    private void cancelTaskOnThisNode(String persistentTaskId, ActionListener<Response> listener) {
+        ReindexDataStreamTask runningTask = getRunningPersistentTaskFromTaskManager(persistentTaskId);
+        if (runningTask == null) {
+            listener.onFailure(
+                new ResourceNotFoundException(
+                    Strings.format(
+                        "Persistent task [{}] is supposed to be running on node [{}], but the task is not found on that node",
+                        persistentTaskId,
+                        clusterService.localNode().getId()
+                    )
+                )
+            );
+        } else {
+            runningTask.markAsCompleted();
```

Comment:
So is this all that's needed to cancel the actual reindex code? Or does there need to be cancel specific logic built into the per-index reindexing code, which will be called here?

Reply:
It sounds like `markAsCompleed` is the wrong way to go. I'm going to change this to use `PersistentTasksService::sendRemoveRequest` instead. And we just need to make sure the parent task is set on all of our transport actions to have the cancellation propagate.